Proxy protocol

wcawijngaards opened the pull request:
This is an implementation of PROXYv2 for NSD.

It can be configured with `proxy-protocol-port: <portnum>` with the port number of the interface on which proxy traffic is handled. The interface can support proxy traffic for UDP, TCP and TLS.

It uses code from Unbound, that implements the proxy protocol specs, in https://github.com/NLnetLabs/unbound/pull/881 .

wcawijngaards committed:
- proxy-protocol, add the proxy protocol generic code proxy_protocol.c and
  proxy_protocol.h.

wcawijngaards committed:
- proxy-protocol, adjust Makefile.in for util/proxy_protocol.c.

wcawijngaards committed:
- proxy-protocol, init proxy protocol write function callbacks.

wcawijngaards committed:
- proxy-protocol, change to store proxy address in struct query.

wcawijngaards committed:
- proxy-protocol, adjust generic header for IPv6 support with ifdef.

wcawijngaards committed:
- proxy-protocol, handle udp proxy header.

wcawijngaards committed:
- proxy-protocol, check size for safety.

wcawijngaards committed:
- proxy-protocol, take care with unaligned reads for safety.

wcawijngaards committed:
- proxy-protocol, verbose error output for header read failure.

wcawijngaards committed:
- proxy-protocol, add unit test for proxy protocol.

wcawijngaards committed:
- proxy-protocol, remove whitespace from test text file.

wcawijngaards committed:
- proxy-protocol, use proxy address for access list block checks.

wcawijngaards committed:
- proxy-protocol, remove unused whitespace.

wcawijngaards committed:
- proxy-protocol, error output easier to read.

wcawijngaards committed:
- proxy-protocol, unit test for allow query interaction with proxy address.

wcawijngaards committed:
- proxy-protocol, unit test for provide-xfr interaction with proxy address.

wcawijngaards committed:
- proxy-protocol, unit test for allow-notify interfaction with proxy address.
  Fix for printout of refusal in log.

wcawijngaards committed:
Merge branch 'master' into proxy-protocol

wcawijngaards committed:
- proxy-protocol, make depend.

wcawijngaards committed:
- proxy-protocol, fix compile.

wcawijngaards committed:
- proxy-protocol, handle tcp proxy header.

wcawijngaards committed:
- proxy-protocol, add zone file to test case.

wcawijngaards committed:
- proxy-protocol, unit test for tcp.

wcawijngaards committed:
- proxy-protocol, remove debug output of log err.

wcawijngaards committed:
Merge branch 'master' into proxy-protocol

wcawijngaards committed:
- proxy-protocol, test setup for TLS.

wcawijngaards committed:
- proxy-protocol, handle tls proxy header.

wcawijngaards committed:
- proxy-protocol, do not log proxy addr if the query is not proxied.

wcawijngaards committed:
- proxy-protocol, unit test for tls.

wcawijngaards committed:
- proxy-protocol, proxy-protocol-port: portno option.

wcawijngaards committed:
Merge branch 'master' into proxy-protocol

wcawijngaards committed:
- proxy-protocol, unit test for dnstap log with proxy-protocol client address.

k0ekk0ek commented on server.c:
```diff
@@ -131,6 +132,16 @@ log_addr(const char* descr,
 #endif
 
+/* header state for the PROXYv2 header (for TCP) */
+enum pp2_header_state_type {
```
Normally, we reserve `_type` for `typedef`'s. Maybe `typedef enum pp2_header_state { ...} pp2_header_state_type` is more consistent?

wcawijngaards replied:
Fixed in https://github.com/NLnetLabs/nsd/pull/281/commits/427739ac972b9395ba5d257e381115ce9ddc19b2 . Used the type name without the `_type` suffix.

k0ekk0ek commented on server.c:
```diff
@@ -131,6 +132,16 @@ log_addr(const char* descr,
+/* header state for the PROXYv2 header (for TCP) */
+enum pp2_header_state_type {
+	/* no header encounter yet */
+	pp2_header_none = 0,
+	/* read the static part of the header */
+	pp2_header_init,
+	/* read the full header */
+	pp2_header_done
```
Perhaps it's better to use uppercase for `enum` labels instead?

wcawijngaards replied:
Because this is not a set of type identifiers, like the enum for like the RR type list is, I think the lowercase is fine. It is also in other places like that. So, I do not see why to change the code style. Thanks for the heads up, though.

k0ekk0ek commented on server.c:
```diff
@@ -3570,6 +3603,88 @@ port_is_zero(
+		 * No need to do anything with addresses. */
+		goto done;
+	}
+	if(header->fam_prot == 0x00) {
```
Why do this check outside the `switch`?

wcawijngaards replied:
This is also in the Unbound proxy protocol code. I think it is because the 0x00 check is for an out of band message, of a healthcheck type, and it does not contain an address. In the switch it deals with the address, so this skips the address management part, to do other commands.

Apart from that, the code could be in the switch, technically. So I would not mind really, either way. I'll keep it like it is now, because of the flow of the algorithm from the comments.

k0ekk0ek commented on options.c:
```diff
@@ -2578,3 +2630,32 @@ resolve_interface_names(struct nsd_options* options)
+	if(ss->ss_family == AF_INET6) {
+		struct sockaddr_in6* a6 = (struct sockaddr_in6*)addr;
+		port = ntohs(a6->sin6_port);
+	} else if(ss->ss_family == AF_INET) {
```
This seems incorrect? As in, protocol can still be something else than AF_INET if INET6 isn't defined.

wcawijngaards replied:
Fixed in https://github.com/NLnetLabs/nsd/pull/281/commits/944aa2bdef5fac1d9b68de800c0f6cb7da1cd4c5 . That checks for the sockaddr family also when INET6 is not defined. The code does not actually have other sockets than AF_INET and AF_INET6 sockets for this part, so this is a depth defense check, that is not really needed, like an assert.

k0ekk0ek commented on util/proxy_protocol.c:
```diff
@@ -0,0 +1,194 @@
+	/* version and command */
+	*buf = (PP2_VERSION << 4) | PP2_CMD_PROXY;
+	buf++;
+	if(af==AF_INET) {
```
This check is inconsistent? We define `INET`, `INET6` and `UNIX` socket types, but we only support `INET` and `INET6` here?

wcawijngaards replied:
This code, util/proxy_protocol.c is from the generic proxy protocol code, that is in Unbound.

The code really only has `INET` and `INET6` socket types that use these routines, so it is safe. But from a defense in depth point, it is possible to have the extra checks. That said, I am fine with either change, of an added check or not.

gthess replied:
Since this code is supposed to be generic I agree with @k0ekk0ek 's observation although the code is not used that way (yet). Turning this into a switch statement where `UNIX` and `default` would return 0 seems like a nice approach.

gthess replied:
Resolved in d8a47b171ca92092af8c33a888487726f72c77b4.

k0ekk0ek commented on util/proxy_protocol.c:
```diff
@@ -0,0 +1,194 @@
+	}
+	/* Check for supported family and protocol */
+	if(header->fam_prot != 0x00 /* AF_UNSPEC|UNSPEC */ &&
+		header->fam_prot != 0x11 /* AF_INET|STREAM */ &&
```
Maybe it's better to define constants for the combinations instead? So for instance `PP2_INET_STREAM` etc? Makes it easier to read.

wcawijngaards replied:
This code, util/proxy_protocol.c is from the generic proxy protocol code, that is in Unbound.

The constants for it as defined in util/proxy_protocol.h. I think it was chosen to display like this to be more readable, than the bit munging that is needed with the constants, because the values align with hex characters. That said, I am fine with this code but also with a use of the defined constants.

gthess replied:
Defined constants should work better since I see that these combinations are also used elsewhere in the code.

gthess replied:
Resolved in d8a47b171ca92092af8c33a888487726f72c77b4.

k0ekk0ek commented on util/proxy_protocol.h:
```diff
@@ -0,0 +1,164 @@
+/**
+ * PROXYv2 address family.
+ */
+enum pp2_af {
```
Maybe specify these definitions originate from the PROXYv2 protocol specification and are passed on the wire?

wcawijngaards replied:
This code, util/proxy_protocol.h is from the generic proxy protocol code, that is in Unbound.

You think that would improve the comment, I would have thought that fairly obvious. But that said, I am okay with the change. More comments can make for more readable and hence better maintained code.

gthess replied:
I see what @k0ekk0ek means and it would help differentiate between these enums and `pp_parse_errors` further down for example. Would something like `PROXYv2 address family (wire value)` work for this comment and the rest here?

gthess replied:
Resolved in d8a47b171ca92092af8c33a888487726f72c77b4. Used `(protocol value)` instead.

k0ekk0ek commented on server.c:
```diff
@@ -3854,10 +3986,136 @@ handle_tcp_reading(int fd, short event, void* arg)
+				/* nothing more to read; header is complete */
+				data->pp2_header_state = pp2_header_done;
+			} else if(data->bytes_transmitted < current_read_size) {
+				received = read(fd,
```
Seems we could benefit from implementing a function to read+handle errors instead of repeating the same? Can be postponed of course, but seems convenient

wcawijngaards replied:
It is in commit https://github.com/NLnetLabs/nsd/pull/281/commits/db581ac35ed36ee59694e967e313f949da676862 . That creates functions for TCP and TLS to read more data and handle an error.

k0ekk0ek reviewed (approved):
Some small remarks about things that stood out. Nothing blocking.

wcawijngaards committed:
- proxy-protocol, review comment, fix enum type name.

wcawijngaards committed:
Merge branch 'master' into proxy-protocol

wcawijngaards committed:
- proxy-protocol, review comment, check family for sockaddr.

wcawijngaards committed:
- proxy-protocol, review comment, implement function to read and handle error.

pettai commented:
Hi,
Any update on when this is going to be merged into the main branch?

k0ekk0ek commented:
Hi @pettai, it's waiting for one more review to complete. Expect this to be merged soon.

pettai commented:
I noted an issue #297 then merging this branch to the latest git version of NSD, so apart from the review, this also needs to be fixed before merging. Thx

wcawijngaards committed:
- proxy-protocol, Fix #297, NSD: handle_tcp_reading: Assertion `buffer_remaining(data->query->packet) > 0' failed.

wcawijngaards commented:
The commit https://github.com/NLnetLabs/nsd/commit/dfcdadb828f6484d82c7219a2990cd142d716885 fixes the issue. The buffer was not reset properly between queries, the commit fixes that.

gthess reviewed (approved):
The PR looks nice!
Just replied on @k0ekk0ek 's suggestions.

gthess committed:
proxy-protocol, review comments:
- more generic switch statement for address families;
- comment the protocol values as such in their definitions;
- less hardcoded values for address family and protocol combinations.

wcawijngaards committed:
proxy-protocol, guard AF_INET6 use with ifdef.

pettai commented:
So it has been done 🎉
